File: src/java/spring/src/main/java/com/benchmark/service/DatabaseService.java
package com.benchmark.service;

import com.benchmark.model.User;
import com.benchmark.model.UserStats;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
public class DatabaseService {
    private final JdbcTemplate jdbcTemplate;

    @Autowired
    public DatabaseService(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    public Optional<User> getUserById(Integer id) {
        if (id == null) {
            return Optional.empty();
        }

        try {
            var userMap = jdbcTemplate.queryForMap(
                "SELECT id, email, first_name, last_name, age, created_at FROM users WHERE id = ?",
                id
            );

            User user = new User();
            user.setId((Integer) userMap.get("id"));
            user.setEmail((String) userMap.get("email"));
            user.setFirstName((String) userMap.get("first_name"));
            user.setLastName((String) userMap.get("last_name"));
            user.setAge((Integer) userMap.get("age"));
            user.setCreatedAt(((java.sql.Timestamp) userMap.get("created_at")).toLocalDateTime());

            return Optional.of(user);
        } catch (org.springframework.dao.EmptyResultDataAccessException e) {
            return Optional.empty();
        }
    }

    public List<UserStats> getUserStats(int days) {
        List<UserStats> stats = new ArrayList<>();

        if (days <= 0 || days > 365) {
            throw new IllegalArgumentException("Days must be between 1 and 365");
        }

        var results = jdbcTemplate.queryForList("""
            -- Normative SQL, see contracts/rest/canonical-payloads.md. The previous
            -- query joined order_items, aggregated quantity*price and ordered without
            -- a tiebreak, so it ran a heavier query than the other implementations and
            -- its rows came back in arbitrary order among equal values.
            SELECT
                u.id AS "userId",
                u.first_name || ' ' || u.last_name AS "userName",
                COUNT(o.id) AS "totalOrders",
                COALESCE(SUM(o.total_amount), 0) AS "totalValue",
                COALESCE(AVG(o.total_amount), 0) AS "averageOrderValue"
            FROM users u
            INNER JOIN orders o ON u.id = o.user_id
                WHERE o.created_at >= NOW() - INTERVAL '1 day' * ?
            GROUP BY u.id, u.first_name, u.last_name
            ORDER BY "totalOrders" DESC, u.id
            LIMIT 100
            """, days);

        for (var row : results) {
            UserStats userStats = new UserStats();
            userStats.setUserId((Integer) row.get("userId"));
            userStats.setUserName((String) row.get("userName"));
            userStats.setTotalOrders(((Number) row.get("totalOrders")).intValue());
            userStats.setTotalValue(((Number) row.get("totalValue")).doubleValue());
            userStats.setAverageOrderValue(((Number) row.get("averageOrderValue")).doubleValue());
            stats.add(userStats);
        }

        return stats;
    }
}

File: src/kotlin/spring/src/main/kotlin/com/benchmark/service/DatabaseService.kt
package com.benchmark.service

import com.benchmark.model.User
import com.benchmark.model.ComplexResult
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class DatabaseService(
    private val jdbcTemplate: JdbcTemplate
) {
    fun getUserById(id: Int): User? {
        val sql = """
            SELECT id, email, first_name AS "firstName", last_name AS "lastName",
                   age, created_at AS "createdAt"
            FROM users
            WHERE id = ?
        """

        val users = jdbcTemplate.query(sql, { rs, _ ->
            User(
                id = rs.getInt("id"),
                email = rs.getString("email"),
                firstName = rs.getString("firstName"),
                lastName = rs.getString("lastName"),
                age = rs.getObject("age") as? Int,
                createdAt = rs.getTimestamp("createdAt").toLocalDateTime()
            )
        }, id)

        return users.firstOrNull()
    }

    fun getComplexQuery(days: Int): List<ComplexResult> {
        val sql = """
            -- Normative SQL, see contracts/rest/canonical-payloads.md. The previous
            -- query joined order_items, aggregated quantity*price and ordered without
            -- a tiebreak, so it ran a heavier query than the other implementations and
            -- its rows came back in arbitrary order among equal values.
            SELECT
                u.id AS "userId",
                u.first_name || ' ' || u.last_name AS "userName",
                COUNT(o.id) AS "totalOrders",
                COALESCE(SUM(o.total_amount), 0) AS "totalValue",
                COALESCE(AVG(o.total_amount), 0) AS "averageOrderValue"
            FROM users u
            INNER JOIN orders o ON u.id = o.user_id
                WHERE o.created_at >= NOW() - INTERVAL '1 day' * ?
            GROUP BY u.id, u.first_name, u.last_name
            ORDER BY "totalOrders" DESC, u.id
            LIMIT 100
        """

        return jdbcTemplate.query(sql, { rs, _ ->
            ComplexResult(
                userId = rs.getInt("userId"),
                userName = rs.getString("userName"),
                totalOrders = rs.getLong("totalOrders"),
                totalValue = rs.getDouble("totalValue"),
                averageOrderValue = rs.getDouble("averageOrderValue")
            )
        }, days)
    }

    fun healthCheck(): Boolean {
        return try {
            jdbcTemplate.queryForObject("SELECT 1", Int::class.java) == 1
        } catch (e: Exception) {
            println("Database health check failed: ${e.message}")
            false
        }
    }
}
